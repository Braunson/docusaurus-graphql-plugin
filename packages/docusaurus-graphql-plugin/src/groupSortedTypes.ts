import {
  GraphQLNamedType,
  isEnumType,
  isInputObjectType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import { GroupedTypes, InterfaceItem, TypeItem } from "./types";

function groupTypes(types: GraphQLNamedType[]): GroupedTypes {
  return types.reduce<GroupedTypes>(
    (acc, type) => {
      if (type.name.startsWith("__")) {
        return acc;
      }

      if (isObjectType(type)) {
        if (type.name === "Query") {
          return {
            ...acc,
            queries: Object.values(type.getFields()).map((query) => ({
              type: query,
            })),
          };
        }

        if (type.name === "Mutation") {
          return {
            ...acc,
            mutations: Object.values(type.getFields()).map((mutation) => ({
              type: mutation,
            })),
          };
        }

        type.getInterfaces().forEach((inter) => {
          acc.interfaces.forEach((otherInter) => {
            if (otherInter.type.name === inter.name) {
              otherInter.implementedBy.push(type);
            }
          });
        });

        return {
          ...acc,
          objects: acc.objects.concat([{ type }]),
        };
      }

      if (isInterfaceType(type)) {
        const interfaceItem: InterfaceItem = {
          type,
          implementedBy: [],
        };

        acc.objects.forEach(({ type: object }) => {
          object.getInterfaces().forEach((otherInterface) => {
            if (otherInterface.name === interfaceItem.type.name) {
              interfaceItem.implementedBy.push(object);
            }
          });
        });

        return {
          ...acc,
          interfaces: acc.interfaces.concat([interfaceItem]),
        };
      }

      if (isEnumType(type)) {
        return {
          ...acc,
          enums: acc.enums.concat([{ type }]),
        };
      }

      if (isUnionType(type)) {
        return {
          ...acc,
          unions: acc.unions.concat([{ type }]),
        };
      }

      if (isInputObjectType(type)) {
        return {
          ...acc,
          inputObjects: acc.inputObjects.concat([{ type }]),
        };
      }

      if (isScalarType(type)) {
        return {
          ...acc,
          scalars: acc.scalars.concat([{ type }]),
        };
      }

      return acc;
    },
    {
      queries: [],
      mutations: [],
      objects: [],
      interfaces: [],
      enums: [],
      unions: [],
      inputObjects: [],
      scalars: [],
    }
  );
}

function sortByName<T extends TypeItem<{ name: string }>>(types: T[]): T[] {
  return types.sort((a, b) => a.type.name.localeCompare(b.type.name));
}

function sortGroupedTypes(groupedTypes: GroupedTypes): GroupedTypes {
  return {
    queries: sortByName(groupedTypes.queries),
    mutations: sortByName(groupedTypes.mutations),
    objects: sortByName(groupedTypes.objects),
    interfaces: sortByName(groupedTypes.interfaces),
    enums: sortByName(groupedTypes.enums),
    unions: sortByName(groupedTypes.unions),
    inputObjects: sortByName(groupedTypes.inputObjects),
    scalars: sortByName(groupedTypes.scalars),
  };
}

export function groupSortedTypes(types: GraphQLNamedType[]): GroupedTypes {
  return sortGroupedTypes(groupTypes(types));
}
