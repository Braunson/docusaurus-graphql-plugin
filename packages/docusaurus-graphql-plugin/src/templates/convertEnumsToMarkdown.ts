import { MarkdownOptions, EnumItem } from "../types";
import { parseMarkdown } from "./parseMarkdown";

export function convertEnumsToMarkdown(
  enums: EnumItem[],
  _: MarkdownOptions
): string {
  const lines: string[] = [];

  enums.forEach(({ type: en }) => {
    lines.push(`## ${en.name}`, `\n\n`);
    lines.push(en.description || "", `\n\n`);

    lines.push(
      `<p style={{ marginBottom: "0.4em" }}><strong>Values</strong></p>`,
      `\n\n`
    );

    lines.push(`<table>`, `\n`);
    lines.push(
      `<thead><tr><th>Value</th><th>Description</th></tr></thead>`,
      `\n`
    );
    lines.push(`<tbody>`, `\n`);

    en.getValues().forEach((value) => {
      lines.push(`<tr>`, `\n`);
      lines.push(`<td>${value.value}</td>`, `\n`);
      lines.push(`<td>`, `\n`);
      lines.push(parseMarkdown(value.description || ""), `\n`);
      lines.push(`</td>`, `\n`);
      lines.push(`</tr>`, `\n`);
    });

    lines.push(`</tbody>`, `\n`);
    lines.push(`</table>`, `\n\n`);
  });

  return lines.join("");
}
