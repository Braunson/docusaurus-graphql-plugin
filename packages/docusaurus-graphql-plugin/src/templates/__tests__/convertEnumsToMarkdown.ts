import { buildSchema } from "graphql";
import { groupSortedTypes } from "../../groupSortedTypes";
import { convertEnumsToMarkdown } from "../convertEnumsToMarkdown";

describe("convertEnumsToMarkdown", () => {
  function getTypePath() {
    return "#";
  }

  it("should render enums", () => {
    const schema = buildSchema(`
      """
      Represents a user's role.
      """
      enum UserRole {
        "User has admin rights."
        Admin
        "User has no particular rights."
        User
      }
    `);
    const { enums } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertEnumsToMarkdown(enums, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });
});
