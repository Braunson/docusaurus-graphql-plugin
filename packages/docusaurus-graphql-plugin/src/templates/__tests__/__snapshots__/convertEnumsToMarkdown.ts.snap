// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertEnumsToMarkdown should render enums 1`] = `
"## UserRole

Represents a user's role.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Values</strong></p>

<table>
<thead><tr><th>Value</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>Admin</td>
<td>
<p>User has admin rights.</p>
</td>
</tr>
<tr>
<td>User</td>
<td>
<p>User has no particular rights.</p>
</td>
</tr>
</tbody>
</table>

"
`;
